src/instruments/src/SD/Pages/Hyd/Hyd.tsx

```diff
@@ -0,0 +1,776 @@
+    x: number,
+    y: number,
+    hydLevelLow: boolean,
+    fireValve: number,
+    pumpDetectLowPressure: number
+    pressure: number
+}
+
+const HydEngPump = ({ system, pumpOn, x, y, hydLevelLow, fireValve, pumpDetectLowPressure, pressure } : HydEngPumpProps) => {
+    const lowPressure = 1450;
+    if (system === 'BLUE') {
+        return (
+            <>
+                <line className={pressure <= lowPressure ? 'AmberLine' : 'GreenLine'} x1={x} y1={y - 32} x2={x} y2={y - 80} />
+                <rect className={pumpDetectLowPressure ? 'AmberLine' : 'GreenLine'} x={x - 16} y={y - 32} width={32} height={32} />
+                <line className={!pumpDetectLowPressure ? 'GreenLine' : 'Hide'} x1={x} y1={y} x2={x} y2={y - 32} />
+                <line className={pumpOn ? 'Hide' : 'AmberLine'} x1={x - 12} y1={y - 16} x2={x + 12} y2={y - 16} />
+                <text id="ELEC-centre" className={pumpDetectLowPressure && pumpOn ? 'RatPtuElec FillAmber' : 'Hide'} x={x} y={y - 16} alignmentBaseline="central">LO</text>
+
+            </>
+        );
+    }
+
+    return (
+        <>
+            <rect className={pumpDetectLowPressure ? 'AmberLine' : 'GreenLine'} x={x - 16} y={y - 80} width={32} height={32} />
+            <line className={!pumpDetectLowPressure ? 'GreenLine' : 'Hide'} x1={x} y1={y} x2={x} y2={y - 80} />
+            <line className={pumpOn ? 'Hide' : 'AmberLine'} x1={x - 12} y1={y - 64} x2={x + 12} y2={y - 64} />
+            <line className={hydLevelLow || !fireValve ? 'AmberLine' : 'GreenLine'} x1={x} y1={y} x2={x} y2={y - 48} />
+            <text
+                id="ELEC-centre"
+                className={
+                    pumpDetectLowPressure && pumpOn ? 'RatPtuElec FillAmber' : 'Hide'
+                }
+                x={x}
+                y={y - 64}
+                alignmentBaseline="central"
+            >
+                LO
+
+            </text>
+
+        </>
+    );
+};
+
+type HydEngValveProps = {
+    system: string,
+    x: number,
+    y: number,
+    fireValve: number,
+    hydLevelLow: boolean
+}
+
+const HydEngValve = ({ system, x, y, fireValve, hydLevelLow } : HydEngValveProps) => {
+    if (system === 'BLUE') {
+        return (
+            <line className={!hydLevelLow ? 'GreenLine' : 'AmberLine'} x1={x} y1={y + 33} x2={x} y2={y} />
+        );
+    }
+
+    return (
+        <>
+            <circle className={fireValve ? 'GreenLine' : 'AmberLine'} cx={x} cy={y + 16} r="16" />
+            <line className={fireValve ? 'GreenLine' : 'Hide'} x1={x} y1={y + 32} x2={x} y2={y} />
+            <line className={fireValve ? 'Hide' : 'AmberLine'} x1={x - 10} y1={y + 16} x2={x + 10} y2={y + 16} />
+        </>
+    );
+};
+
+type HydReservoirProps = {
+    system: string,
+    x: number,
+    y: number,
+    fluidLevel: number,
+    setHydLevel: React.RefCallback<
+    Boolean>
+}
+
+const HydReservoir = ({ system, x, y, fluidLevel, setHydLevel } : HydReservoirProps) => {
+    const levels = [
```

Comment:
Since this is constant, it should be outside the function so we don't create unnecessary objects every render.

Reply:
I'd like to point out that unless you measured this, your statement is a Schrödinger's cat situation. This is _very much_ depending on the trickery the JS engine performs. For example, it _might_ notice that the array is never modified and therefore optimise this to a single representation in memory. Similarly, when moving this outside of the function it _might_ incur the additional scope lookup overhead or it _might_ be inlined into the function where it _might_ have a single memory representation.

Reply:
That kind of optimization does not really happen in JS engines. The dynamic nature of the language guarantees that *anything* could be modified at *any time* by *anything*.

The engine is not going to inline something declared outside the scope because it has no way to know if a dynamic lookup, an `eval` call or something else will go access/modify that array.

Without measuring,, it's worth noting that modern V8 versions does not optimize this scenario, for example.

Reply:
In the tests I performed, creating new objects often (via `new` or object literals) is very costly every few seconds. I have not figured out how to get the GT debugger to track allocations yet though.

Reply:
Could you provide some reference(s) on these statement: 

> That kind of optimization does not really happen in JS engines.

> Without measuring,, it's worth noting that modern V8 versions do not optimize this scenario, for example.

It's hard to wrap my head around the ability of these engines to compile to bytecode, and for objects created to be transformed into "hidden classes" for which it can then perform inline caching.